### ISSUE TITLE
No files found to build ObjectList considered an error

### ISSUE THREAD

**Issue by missmah** [fastbuild/fastbuild#216]
The situation where FastBuild emits 'No files found to build' for an ObjectList _appears_ to be considered an error. I question whether this in-fact should be an error or not.

We have a case where we want to build certain .cpp files only in internal (non-customer-facing builds). 

**So, we have some code that looks like this:**

```
//------------------------------------------------------------------------------
// Detailed Configurations
//------------------------------------------------------------------------------
.Internal_Build_Configurations = { 'Debug', 'Development', 'Release' }
.External_Build_Configurations = { 'Shipping' }

.Build_Configurations = { .Internal_Build_Configurations, .External_Build_Configurations }

.UnitTestCppFilePattern = '*UnitTests*.cpp'

.ConfigForAllBuilds =
[
	.BuildAudienceType	          = 'EXTERNAL'
	.UnityFilePatternExclusions = .UnitTestCppFilePattern	// Exclude Unit-Tests by Default
	.ExtraFilePatternInclusions = ''
]

// Internal Builds => Non-Shipping Builds
.ConfigForInternalBuilds =
[
	Using( .ConfigForAllBuilds )

	.BuildAudienceType			= 'INTERNAL'
	.ExtraFilePatternInclusions + .UnitTestCppFilePattern	// Include Unit-Tests for Internal Builds
]

.ConfigForAllDebugBuilds =
[
	Using( .ConfigForInternalBuilds )
]

.ConfigForAllDevelopmentBuilds =
[
	Using( .ConfigForInternalBuilds )
]

.ConfigForAllReleaseBuilds =
[
	Using( .ConfigForInternalBuilds )
]

// Shipping Builds are EXTERNAL Builds
.ConfigForAllShippingBuilds =
[
	Using( .ConfigForAllBuilds )
]
```

**Our Unity code looks like this:**

    // Unity
    //--------------------------------------------------------------------------
    {
	Using( .ConfigForAllBuilds )

        // Common options
        .UnityInputPath             = '$ProjectPath$\'
        .UnityOutputPath            = '$OutputBase$\Unity\$ProjectPath$\'
	.UnityInputExcludePattern	= .UnityFilePatternExclusions
       // ...
    }

**We then have some per-configuration code that looks like this:**

```
// Compile Extra-Files into Object Files
ObjectList( '$ProjectName$-ExtraFiles-OBJ-$Platform$-$Config$' )
{
	.CompilerOptions			+ .ConfigurationCompilerOptions
							+ .ConfigurationPostPCHCompilerOptions

    // Input (Unity)
	.CompilerInputPath			= '$ProjectPath$\'
	.CompilerInputPattern		= .ExtraFilePatternInclusions

    // Output
    .CompilerOutputPath         = '$OutputBase$\$ProjectName$\'
}

```
In theory, this would work, and exclude *UnitTests*.cpp from Unity builds, but re-include *UnitTests*.cpp* into ExtraFiles builds, except during customer-facing shipping builds, where those files would not be re-included. However, FastBuild emits 'No files found to build' (which makes sense), and then appears to consider the build failed (which doesn't make sense).

This seems to be a bug, as this kind of behavior seems completely sane. For example, Visual Studio allows you to "Exclude from Build" any file, w.r.t. any combination of Platform and Configuration.

Is there a good reason I'm not seeing why FastBuild considers this case to be an error? Is there a better way to accomplish this intended behavior of excluding certain files or file patterns from certain build configurations?

Thanks!

**Comment by ffulin** [comment]
Currently, FASTBuild considers an entirely empty ObjectList or Library as an error. The reasoning behind this is two-fold:
a) most librarians fail if you try to create a library with no inputs
b) specifying inputs in such a way that doesn't find anything to compile is usually an error, so FASTBuild tried to fail as early as possible to make it more obvious

It's a pretty common use-case to have code you only want to compile in certain configurations, but usually that's done just by excluding files within an existing library. In the case of tests, most users tend to build seperate exes and simply don't build those in configs that don't need them.

If I've understood your example well, then the simplest options to get what you have working would seem to be:

a) Compile the test files in the same ObjectList. If you can compile your tests with the same compiler options as the rest of the library, then you can simply add the extra test objects into the same ObjectList you use to compile the Unity (or exclude them only from the 'Shipping' config.

b) Add a dummy file. If you can't merge them as above because they need unique command line options for the compiler, then adding an empty dummy file to your 'ExtraFiles-OBJ' ObjectList which is always compiled in 'Shipping' would avoid the "no files found issue".

Failing either of those, the other options require a change the FASTBuild:
a) Remove the warning and just let things fail as they may. This would be a last resort as the reason for the behaviour as is still seems valid to me.
b) We could add an option to ObjectList to allow the "empty" behaviour (.CompilerIgnoreNoFilesFound)

**Comment by missmah** [comment]
I actually implemented the dummy option last night - though I really wished for the option where I could pass a boolean about ignoring things.I didn't fully realize option A was a possibility; but, in retrospect it makes sense and seems like the right way to go for this particular use case (I don't yet need different build/link options for the unit tests)Thanks for the thorough reply!                                                                                                                                                                                                                From: ffulinSent: Thursday, March 16, 2017 23:55To: fastbuild/fastbuildReply To: fastbuild/fastbuildCc: Layla; AuthorSubject: Re: [fastbuild/fastbuild] No files found to build ObjectList considered an error (#216)Currently, FASTBuild considers an entirely empty ObjectList or Library as an error. The reasoning behind this is two-fold:
a) most librarians fail if you try to create a library with no inputs
b) specifying inputs in such a way that doesn't find anything to compile is usually an error, so FASTBuild tried to fail as early as possible to make it more obvious
It's a pretty common use-case to have code you only want to compile in certain configurations, but usually that's done just by excluding files within an existing library. In the case of tests, most users tend to build seperate exes and simply don't build those in configs that don't need them.
If I've understood your example well, then the simplest options to get what you have working would seem to be:
a) Compile the test files in the same ObjectList. If you can compile your tests with the same compiler options as the rest of the library, then you can simply add the extra test objects into the same ObjectList you use to compile the Unity (or exclude them only from the 'Shipping' config.
b) Add a dummy file. If you can't merge them as above because they need unique command line options for the compiler, then adding an empty dummy file to your 'ExtraFiles-OBJ' ObjectList which is always compiled in 'Shipping' would avoid the "no files found issue".
Failing either of those, the other options require a change the FASTBuild:
a) Remove the warning and just let things fail as they may. This would be a last resort as the reason for the behaviour as is still seems valid to me.
b) We could add an option to ObjectList to allow the "empty" behaviour (.CompilerIgnoreNoFilesFound)

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/fastbuild/fastbuild","title":"fastbuild/fastbuild","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/fastbuild/fastbuild"}},"updates":{"snippets":[{"icon":"PERSON","message":"@ffulin in #216: Currently, FASTBuild considers an entirely empty ObjectList or Library as an error. The reasoning behind this is two-fold:\r\na) most librarians fail if you try to create a library with no inputs\r\nb) specifying inputs in such a way that doesn't find anything to compile is usually an error, so FASTBuild tried to fail as early as possible to make it more obvious\r\n\r\nIt's a pretty common use-case to have code you only want to compile in certain configurations, but usually that's done just by excluding files within an existing library. In the case of tests, most users tend to build seperate exes and simply don't build those in configs that don't need them.\r\n\r\nIf I've understood your example well, then the simplest options to get what you have working would seem to be:\r\n\r\na) Compile the test files in the same ObjectList. If you can compile your tests with the same compiler options as the rest of the library, then you can simply add the extra test objects into the same ObjectList you use to compile the Unity (or exclude them only from the 'Shipping' config.\r\n\r\nb) Add a dummy file. If you can't merge them as above because they need unique command line options for the compiler, then adding an empty dummy file to your 'ExtraFiles-OBJ' ObjectList which is always compiled in 'Shipping' would avoid the \"no files found issue\".\r\n\r\nFailing either of those, the other options require a change the FASTBuild:\r\na) Remove the warning and just let things fail as they may. This would be a last resort as the reason for the behaviour as is still seems valid to me.\r\nb) We could add an option to ObjectList to allow the \"empty\" behaviour (.CompilerIgnoreNoFilesFound)"}],"action":{"name":"View Issue","url":"https://github.com/fastbuild/fastbuild/issues/216#issuecomment-287260577"}}}

**Comment by gkoreman** [comment]
Layla, is this a new issue for you in 0.93?

Yesterday I attempted to upgrade our build to use FBuild 0.93, however we are using the UE4 integration found here: https://github.com/ClxS/FASTBuild-UE4

This integration uses a somewhat shady workaround to play nice with the UE4 build system. Instead of building static libraries by compiling source and sending the obj files to a librarian, it actually does two passes - it builds obj files with an ObjectList and then it passes the object files to a Library node. For the Library node's compiler it used 'Null', which is undefined. In 0.92 this worked because there were no files to compile, however with 0.93 the reflection system validates the compiler regardless of whether or not any files need to be compiled.

I was considering solving this by:
A) Define a valid compiler (however if Layla's issue is new to 0.93 this still may not work)
B) Allow Library to ignore .Compiler if there are no compiler inputs
C) ???

**Comment by missmah** [comment]
@gkoreman I don't think it's new in 0.93 - we added the exclusion of UnitTests from external/shipping builds in 0.93, so I don't have a reference point for before that on this feature. But we also build Unreal with FastBuild, so I'll have to look at what we're doing. I know things changed dramatically in UE 4.15, and we had to start parsing .response files... I think we were using ObjectLists and DLL nodes instead of library nodes, but I'm not sure that makes much of a difference?

**Comment by ffulin** [comment]
Nothing has (intentionally) changed in v0.93 with respect to this error.

However, that is subtle difference in behaviour between ObjectList and Library. Library can be invoked without any input object files (because it's valid to make a library my merging other libraries). It's might be that somehow your switch from Library to ObjectList is a factor.

**Comment by missmah** [comment]
@gkoreman were you referring to these lines?:

//We do not provide file inputs (we use separate ObjectLists) for simplicity, so we do not need to specify an actually defined compiler.
sb.AppendFormat("\t.Compiler\t\t = 'Null'\n");

It looks like we solved this by defining a valid compiler - I'm guessing you fixed this by now (sorry I've been so slow to get back to this!); but, if not, I can share our code with you.

**Comment by gkoreman** [comment]
For the time being I rolled us back to 0.92. Yes, it's those lines that were the issue. We haven't fixed this yet, but defining a compiler sounds like the path of least resistance. I'll probably go that route when attempting the upgrade again.

**Comment by missmah** [comment]
I decided to share our code preemptively, in-case it helps you (or others) port to 0.93 + 4.15. :)

https://github.com/missmah/FASTBuild-UE4

**Comment by Shelim** [comment]
Can you add mentioned option  `.CompilerIgnoreNoFilesFound`? I need to compile possibly empty directory into distinct path which will be linked together outside FastBuild. There is no way to force compiler to accept it unless void empty file is provided...

**Comment by ffulin** [comment]
A version containing this new option has now been released (v0.96).
